### ISSUE TITLE
Observing difference in Access list for BigQuery datasets

### ISSUE THREAD

**Issue by shj-18** [terraform-google-modules/terraform-google-bigquery#250]
### TL;DR

We observed a change in the access list of newly created datasets while using Terraform module v5.4.2. This change was not observed in v5.4.1.

This behavior seems to have been mitigated by the patch merged in late 2022. https://github.com/terraform-google-modules/terraform-google-bigquery/pull/157

But we are observing the same behavior as before the patch.  

1. Should we be experiencing the behavior that was observed before the patch at present, given that the bug was fixed in late 2022 ?
2. Can you also explain what made the ACL to change between v5.4.1 and v5.4.2 ?

### Terraform Configuration

```hcl
source = "terraform-google-modules/bigquery/google"
version = "~5.4.2"
```

![image](https://github.com/terraform-google-modules/terraform-google-bigquery/assets/144942179/a52d196c-f371-4786-b299-fa60385c4467)

**Comment by g-awmalik** [comment]
I'm unable reproduce this issue locally. I've tried a sample configuration closely resembling what you've provided here with both v5.4.1, v5.4.2 and the latest version. None of them provision access to project Viewers as Big Query Data Viewer at the dataset level.

To further investigate this, can you:
- share the output of `terraform version` for the root config.
- share the complete configuration that reproduces the issue for you.

**Comment by shj-18** [comment]
- Terraform version v1.0.0

**locals.tf** 

locals {

  dataset_writer = [

    "roles/bigquery.dataOwner"

  ]

  dataset_reader = [

    "roles/bigquery.dataViewer"

  ]

  table_permissions_write = [

    "roles/bigquery.dataOwner"

  ]

  table_permissions_read = [

    "roles/bigquery.dataViewer"

  ]

 

  dataset_write_accounts = var.dataset_write_accounts != null ? var.dataset_write_accounts : []

  ##combine the dataset write user and roles

  combined_dataset_write_accounts = flatten([

    for user_value in var.dataset_write_accounts : [

      for role_value in local.dataset_writer : {

        user : lower(user_value)

        role : role_value

      }

    ]

  ])

 

   # if the value of var.dataset_read_accounts null then

  # provide an empty list

 

  dataset_read_accounts = var.dataset_read_accounts != null ? var.dataset_read_accounts : []

  ##combine the dataset write user and roles

  combined_dataset_read_accounts = flatten([

    for user_value in var.dataset_read_accounts : [

      for role_value in local.dataset_reader : {

        user : lower(user_value)

        role : role_value

      }

    ]

  ])

 

  table_write_accounts = var.table_write_accounts != null ? var.table_write_accounts : []

  combined_table_write_accounts = flatten([

    for table_write_account_entry in local.table_write_accounts : [

      for table_write_user_value in table_write_account_entry.users :[

        for table_write_user_role_value in local.table_permissions_write : {

          table_id: table_write_account_entry.table_id

          user : table_write_user_value

          role : table_write_user_role_value

        }

      ]

    ]

  ])

 

  # if the value of var.dataset_read_accounts null then

  # provide an empty list

  table_read_accounts = var.table_read_accounts != null ? var.table_read_accounts : []

  combined_table_read_accounts = flatten([

    for table_read_account_entry in local.table_read_accounts : [

      for table_read_user_value in table_read_account_entry.users :[

        for table_read_role_value in local.table_permissions_read : {

          table_id: table_read_account_entry.table_id

          user : table_read_user_value

          role : table_read_role_value

        }

      ]

    ]

  ])

}

resource "google_bigquery_dataset_iam_member" "google_dataset_iam_member_write_assignment" {

  count      = local.combined_dataset_write_accounts != null ? length(local.combined_dataset_write_accounts) : 0

  project    = var.project_id

  dataset_id = module.bigquery.bigquery_dataset.dataset_id

  role       = local.combined_dataset_write_accounts[count.index].role

  member     = local.combined_dataset_write_accounts[count.index].user

}

 

resource "google_bigquery_dataset_iam_member" "google_dataset_iam_member_read_assignment" {

  count      = local.combined_dataset_read_accounts != null ? length(local.combined_dataset_read_accounts) : 0

  project    = var.project_id

  dataset_id = module.bigquery.bigquery_dataset.dataset_id

  role       = local.combined_dataset_read_accounts[count.index].role

  member     = local.combined_dataset_read_accounts[count.index].user

}

 

 

resource "google_bigquery_table_iam_member" "google_table_iam_member_write_assignment" {

  depends_on = [

    module.bigquery.table_ids

  ]

  count      = local.combined_table_write_accounts != null ? length(local.combined_table_write_accounts) : 0

  project    = var.project_id

  dataset_id = module.bigquery.bigquery_dataset.dataset_id

  table_id   = local.combined_table_write_accounts[count.index].table_id

  role       = local.combined_table_write_accounts[count.index].role

  member     = local.combined_table_write_accounts[count.index].user

}

 

 

resource "google_bigquery_table_iam_member" "google_table_iam_member_read_assignment" {

  depends_on = [

    module.bigquery.table_ids

  ]

  count      = local.combined_table_read_accounts != null ? length(local.combined_table_read_accounts) : 0

  project    = var.project_id

  dataset_id = module.bigquery.bigquery_dataset.dataset_id

  table_id   = local.combined_table_read_accounts[count.index].table_id

  role       = local.combined_table_read_accounts[count.index].role

  member     = local.combined_table_read_accounts[count.index].user

}


**main.tf** 

module "bigquery" {

  source  = "terraform-google-modules/bigquery/google"

  version = "5.4.1"

 

  dataset_id                  = local.dataset_id

  dataset_name                = local.dataset_name

  description                 = var.description

  project_id                  = var.project_id

  location                    = var.location

  default_table_expiration_ms = var.default_table_expiration_ms

 

  tables                     = var.tables

  encryption_key             = var.encryption_key

  dataset_labels             = local.dataset_labels

  delete_contents_on_destroy = var.delete_contents_on_destroy

  external_tables            = var.external_tables

  views                      = var.views

}

**Comment by g-awmalik** [comment]
I noticed this block:

```
resource "google_bigquery_dataset_iam_member" "google_dataset_iam_member_read_assignment" {
  count      = local.combined_dataset_read_accounts != null ? length(local.combined_dataset_read_accounts) : 0
  project    = var.project_id
  dataset_id = module.bigquery.bigquery_dataset.dataset_id
  role       = local.combined_dataset_read_accounts[count.index].role
  member     = local.combined_dataset_read_accounts[count.index].user
}
```

which is primarily responsible for setting the `roles/bigquery.dataViewer` role on the dataset. There are no variables configs present here but depending on what you set in `var.dataset_read_accounts` will drive the dataset reader permissions for the accounts.

**Comment by shj-18** [comment]
Pasting the version details.
```
terraform {

  required_providers {

    google = {

      source  = "hashicorp/google"

      version = "~> 4.60.2"

    }


    google-beta = {

      source  = "hashicorp/google-beta"

      version = "~> 4.60.2"

    }

  }

}
```
